### PR TITLE
feat: Support resource extensions

### PR DESCRIPTION
**Pull request by glb**
## What did you implement

The merge logic for extending resources with the existing framework using `resources.Resources` has some surprising behaviours.

This introduces a more formalized definition for extending resources, where `Metadata` and `Properties` will be shallowly merged (the values for existing keys will be replaced instead of merged), the `DependsOn` list will be added to, and the values for `CreationPolicy`, `DeletionPolicy`, `UpdatePolicy`, and `UpdateReplacePolicy` will be replaced if they exist in the extension.

Extension of other resource attributes is not supported at this time.

Related to but does not completely address the requirements in #6575, as fully complying with those requirements would be a breaking change.

Related to #7338.

## How can we verify it

1. Copy `serverless.yml` below to disk.

2. Run `serverless package`.

3. Examine `.serverless/cloudformation-template-update-stack.json`. Validate that the `SampleLogGroup` resource has `RetentionInDays` set to `"30"`.

See also the automated tests that validate the full behaviour.

<details>
<summary>serverless.yml</summary>

```yml
service: test

provider:
  name: aws
  runtime: go1.x
  stage: dev

functions:
  sample:
    handler: bin/sample
    events:
      - http:
          method: post
          path: /sample

resources:
  extensions:
    SampleLogGroup:
      Properties:
        RetentionInDays: '30'

package:
  exclude:
    - ./**
  include:
    - ./bin/**
```

</details>

## Todos

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/package/lib/mergeCustomProviderResources.js [hunk 1]
```diff
@@ -12,11 +12,72 @@ module.exports = {
...
+
     _.merge(
       this.serverless.service.provider.compiledCloudFormationTemplate,
       this.serverless.service.resources
     );
 
+    if (extensions) {
+      const template = this.serverless.service.provider.compiledCloudFormationTemplate;
+
+      _.forOwn(extensions, (resourceDefinition, resourceName) => {
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/mergeCustomProviderResources.js]: Let's resort to lodash, only if we do not have native counterpart. In that case it'll be nicer to construct it with `for..of` as:

```javascript
for (const [resourceName, resourceDefinition] of _.entries(extensions)) {
...
```

We'd be able to replace `_.entries` with `Object.entries` once we drop support for Node.js v6 (and that should happen shortly)

**Comment by glb** [comment]
@medikoo I've updated to use `for..of` as requested; I did it as a separate commit so it would be easier to review.

@pmuens I'm going to assume your question was for @medikoo :)

**Comment by medikoo** [comment]
> Do we also plan to support other cloud providers as well?

I think it's a CloudFormation specific thing. Not sure if other providers have that, and if, whether it's still named _resources_ (?)

Technically currently `resources` is top-level, although in my understanding it was designed just for AWS (is that correct?)

If we agree that all top level should be assumed as provider-agnostic, then I would probably move it to `provider.resources` or even `provider.cloudFormationResources`, to indicate clearly we're handling here CloudFormation template

@pmuens What are your thoughts?

**Comment by glb** [comment]
@medikoo / @pmuens I'm not super-comfortable with the direction this conversation is taking ... this PR came about because [@medikoo suggested that `provider.apiGateway.deployment.dependsOn` was the wrong direction](https://github.com/serverless/serverless/issues/7338#issuecomment-586754203):

> (I wouldn't propagate it to provider namespace, as we don't keep there CF logic) 

If you're talking about a future change to remove `resources` as part of the final resolution of #6575, then could I suggest that maybe that's separate from whether this PR is a valid intermediate step? 

Personally I liked the direction that this was going in, even though it was different from what I'd originally proposed, but ultimately my (selfish) goal is to be able to reliably add something to the `AWS::ApiGateway::Deployment`'s `DependsOn` attribute, ideally soonish. Currently I can't do that, and actually I'm reminded that unless https://github.com/jacob-meacham/serverless-plugin-bind-deployment-id/pull/26 gets merged I still won't be able to, even if this change goes in.

**Comment by medikoo** [comment]
@glb This PR is intended to be purely about introduction of `resources.extensions`, we're not expecting any changes to `resources` here.

What I'm discussing with @pmuens is wheter it's not better to provide _extensions_ at `provider.cloudFormationResources.extensions` instead of placing them at `resources.extensions` (that's all)

Simply to coin right config schema from a start.

We will do changes to `resources` in other PR, and that'll be a breaking change to be published with v2

**Comment by glb** [comment]
Thanks @medikoo ... I think you had a very good point when you suggested that `extensions` should be placed with the existing `resources`, because in a given service they are going to be closely related. 

Personally I put `provider` near the top of my `serverless.yml` and `resources` near the end; perhaps I'm alone in this but I don't think so. Putting the `extensions` under `provider` would make these definitions pretty much as far apart as possible in the file.

**Comment by medikoo** [comment]
As I inspected history of `resources` config. It was first proposed as `resources.aws` and then `aws` was stripped here: https://github.com/serverless/serverless/commit/2c20687793296bc7adcbfb20993782b3ae8c4fe1 

I assume that intention was that `resources` will be a provider agnostic option, which will naturally resort to CloudFormation resources in case of AWS, and eventually some other concept in case of other providers.

It's probably ok to not change that

> How would an extension work if we implement this for another provider (e.g. Azure)? Is an extension provider independent or should it only be used for AWS?

@pmuens I think idea should be same as in case of AWS, that `resources.extensions` should be about extending resources created by the framework.

General idea is to replace `resources` with something as `resources.additions` and `resources.extensions`, to ensure bug free handling. History shown that extending resources as created by Framework demands extra care to avoid issues (we cannot just deeply merge it as it's the case now, as sometimes deep property merge produces invalid structures, and sometimes user is not be aware that framework genereated resource will be overriden with his setup).

This distinction will allow to fix such issues.

In v1 we can add `resources.extensions` and with v2 we would be able to add `resources.addtions` and do not support any setup directly on `resources`

**Comment by pmuens** [comment]
> Technically currently resources is top-level, although in my understanding it was designed just for AWS (is that correct?)

The `resources` section was introduced to be a way for all providers to bring in additional resource one needs to spin up a whole service.

> If we agree that all top level should be assumed as provider-agnostic, then I would probably move it to provider.resources or even provider.cloudFormationResources, to indicate clearly we're handling here CloudFormation template

That sounds reasonable, however I agree with @glb that the `resources` are usually defined at the bottom and the `provider` config is defined at the top of the `serverless.yml` file.

IMHO it's not a big deal to introduce the `extensions` config at the `resources` level. I just wanted to ensure that we'll implement this in a way that it's provider agnostic (maybe we need to re-use this in other provider plugins as well).

**Comment by medikoo** [comment]
@pmuens can you give it a second look?

**Pull request merged**